src/blockcrawler/evm/rpc.py

```diff
@@ -195,6 +195,118 @@ async def get_block(self, block_num: HexInt, full_transactions: bool = False) ->
         )
         return block
 
+    async def get_block_by_timestamp(self, timestamp: HexInt) -> EvmBlock:
+        """Get block number by timestamp.
+
+        :param timestamp: The timestamp of the block you wish to get.
+        :returns: The block which timestamp is equal to given timestamp
+            or the nearest block if there's no block with the exact timestamp.
+        """
+
+        return await self.__get_block_by_timestamp(timestamp)
+
+    async def __get_block_by_timestamp(
+        self,
+        timestamp: HexInt,
+        left_block: Optional[EvmBlock] = None,
+        right_block: Optional[EvmBlock] = None,
+    ) -> EvmBlock:
+        """Recursively get block number by timestamp until exact or nearest match.
+
+        :param timestamp: The timestamp of the block you wish to get.
+        :param left_block: The left bound block limit of binary search.
+        :param right_block: The right bound block limit of binary search.
+        :returns: The block which timestamp is equal to given timestamp
+            or the nearest block if there's no block with the exact timestamp.
+        """
+
+        # Get block by timestamp binary search source: https://ethereum.stackexchange.com/a/127720
+
+        if not left_block or not right_block:
+            # Ethereum's Paris Network Upgrade (a.k.a. the Merge!)
+            #   Block Number: 15537394
+            #   Timestamp: 1663224179
+            the_merge_timestamp = HexInt(1663224179)
+            the_merge_block_number = HexInt(15537394)
+
+            if timestamp == the_merge_timestamp:
+                # Return known value.
+                return await self.get_block(the_merge_block_number)
+
+            right_block = await self.get_block(await self.get_block_number())
+
+            # Check if timestamp is later than the network upgrade event
+            #   where each block's duration are only either 12 or 24 seconds.
+            # 24 seconds block duration is only occurring for about 2% of the time
+            if timestamp > the_merge_timestamp:
+                min_block_duration = HexInt(12)
+
+                estimated_adjustment = (right_block.timestamp - timestamp) / min_block_duration
+                estimated_block_number = right_block.number - estimated_adjustment
+
+                # Set left bound block to the estimated block number.
+                # Estimated block is always less than or equal to the expected block.
+                left_block = await self.get_block(
+                    min(max(estimated_block_number, the_merge_block_number), right_block.number)
+                )
+
```

Comment:
This seems very specific to Ethereum. Block Crawler is blockchain agnostic. It must work with all blockchains.

Reply:
Addressed.